feat(auth): Added keycloak as a social login provider

pa1narendra opened the pull request:
This pull request is a feature request to add keycloak as a social login provider as it is used by many companies and this is really really helpful for us, @akhilnarang can you please review this and accept this if there are no changes , this will pass all the pre-commit tests and I am sure that everything is well set. Please review this and accept as soon as possible. Previous pr accidentally closed due to force push which I had done to make sure that the commit should pass all the tests. And I am really sorry for that. Can you please review and accept this pull request.

It is also requested by people previously in the issues #24563 and in my previous PR

pa1narendra committed:
feat(auth): added keycloak as a social login provider

clayrisser commented:
@akhilnarang is this something you can review and accept soon?

akhilnarang commented on frappe/integrations/doctype/social_login_key/social_login_key.py:
```diff
@@ -80,6 +88,8 @@ def validate(self):
 			frappe.throw(
 				_("Please enter Client Secret before social login is enabled"), exc=ClientSecretNotSetError
 			)
+		if self.social_login_provider == "Keycloak":
```
Seems fine, but this block shouldn't be needed as per my understanding. Setting the path in the provider configuration should be enough

pa1narendra replied:
No, we need that. Not everyone will have the same URL as their base URL and the endpoint. And also API endpoint requires base URL in it, because API endpoint is not editable , we are making it (editable) according to the base URL.

akhilnarang replied:
The library automatically adds the base URL, that's why it shouldn't be required. That's how it works for other providers like Frappe.

pa1narendra replied:
@akhilnarang I'm sorry to say this, but the library does NOT add the base URL. If you would test this yourself you would be able to confirm this. It is well documented and many others have faced this EXACT same issue. The base URL MUST be added in order for it to work with Keycloak. Either frappe needs to fix this fundamental issue or else this code is required. 

If you do not believe me, please verify it yourself. You can also see how other people need to manually add the base url in the following issue.

https://github.com/frappe/erpnext/issues/35542#issuecomment-1657126920

akhilnarang replied:
@pa1narendra: https://github.com/maxcountryman/rauth/blob/master/rauth/session.py#L38-L41

https://github.com/frappe/frappe/blob/develop/frappe/utils/oauth.py#L151

What behaviour do you face if you just add the path to the config like how all others are configured, and don't set a base URL there, and drop this block?
Currently the API endpoint you've set in the config and the one you're setting here seem to be slightly different.
If the relative URL is correct and the base URL is the same, this shouldn't be an issue.

I'll try to setup a test keycloak instance to check this.

akhilnarang replied:
Read up some more: so basically this is because the realm is dynamic?

Could you share some complete URLs for the API endpoint, authorize URL, access token URL

pa1narendra replied:
If I remove the base URL from the API Endpoint and the block of the code, this is the result. 

![2024-09-11-141807_screenshot](https://github.com/user-attachments/assets/8bee98a8-faff-4bbf-8146-15fb0004b848)


![image](https://github.com/user-attachments/assets/c3682834-f264-46f3-8887-787aee817a17)


What you said is working fine with every other provider like you said , but it is not working with Keycloak. Not just me , everyone who uses Keycloak faced the exact same issue. According to the documentation, the base URL is not required but in the case with keycloak, the API Endpoint must be an absolute URL in order to work, if we remove the base URL from the API Endpoint it is giving a email not verified error even though the email is verified. And the reason it is not working because the realm is dynamic like you said.

https://github.com/frappe/erpnext/issues/35542#issuecomment-1575279869


And the Config I am using for this is 

API Endpoint :          http://localhost:8080/realms/masterl/protocol/openid-connect/userinfo
Authorize URL :          /protocol/openid-connect/auth
Access Token URL :         /protocol/openid-connect/token

akhilnarang replied:
And for this your base URL is just `http://localhost:8080`?

Or `http://localhost:8000/realms/masterl`

pa1narendra replied:
this is my base URL `http://localhost:8080/realms/master`

akhilnarang replied:
Thanks for the information, will experiment a bit.

pa1narendra replied:
Sure but can you do it quick , we really need this to be done fast.

pa1narendra replied:
@akhilnarang  Can you please review and accept it.

clayrisser replied:
@akhilnarang any updates on this pull request?

akhilnarang replied:
Haven't gotten time to check some things that I wanted to yet.

clayrisser replied:
Ok, let me know if you need any help setting up keycloak to test it.

akhilnarang replied:
@pa1narendra I've pushed a commit, can you test that, with the base URL set like this: `http://localhost:8080/realms/{realm_name}`
![image](https://github.com/user-attachments/assets/8c7d7db2-2506-45aa-96bc-d5fe12ee80a5)

pa1narendra replied:
@akhilnarang  Is that working after you made those changes ???

akhilnarang replied:
@pa1narendra it'll work now, do check.

I moved the URL setting to another place where it seemed more appropriate, and added a comment explaining why its needed.

pa1narendra replied:
@akhilnarang  Can you test it once again ? I tried and it didn't worked for me. Or is there anything that I want to know while configure and running ?, I am getting the same Email not verified error and also the API Endpoint URL is not updated on save. Can you tell me what changes I have to make it work ?



![2024-09-19-170052_screenshot](https://github.com/user-attachments/assets/26082c10-f196-4a97-ab65-9775e2ce1e6c)

akhilnarang replied:
![image](https://github.com/user-attachments/assets/6d0ebbf8-eef3-42ea-b68e-145cae0fa4d5)

This was my setup and it worked, weird, I was able to login.

I'll try again.

pa1narendra replied:
Is the API Endpoint getting updated on save ??

akhilnarang replied:
No, intended.

pa1narendra replied:
My bad , it is working somehow, I actually forgot to check the email verified and after I checked , it is perfectly working , I can be able to login. Thank you @akhilnarang .

akhilnarang committed:
refactor(keycloak): set API endpoint while getting list of oauth providers

urljoin() just adds the base URL to the relative path, here our base URL has an extra route in the path which isn't added in.

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

clayrisser commented:
@akhilnarang thank you!
